### ISSUE TITLE
No longer use type classes for encoding/decoding SQL rows?

### ISSUE THREAD

**Issue by JordanMartinez** [rightfold/purescript-postgresql-client#40]
After reading through [Thoughts on Type Class Codecs](http://code.slipthrough.net/2018/03/13/thoughts-on-typeclass-codecs/), should we follow suit for encoding/decoding values to/from SQL rows? For example, [`purescript-codec`](https://pursuit.purescript.org/packages/purescript-codec/3.0.0) and its application for JSON, [`purescript-codec-argonaut`](https://pursuit.purescript.org/packages/purescript-codec-argonaut/7.0.2)

**Comment by paluh** [comment]
Hi @JordanMartinez 

Thanks for writing this question. I think I fully understand your concerns as I often feel like a member of "[scrap your typeclasses](http://www.haskellforall.com/2012/05/scrap-your-type-classes.html)" camp ;-)

However I think that in the case of this library we have a bit simpler scenario than in case of arbitrary ADT encoding as described in this interesting @garyb post. Usually types which we dealing with are simple and this encoding / decoding + standard `newtype` wrapping strategy seems to be working fine in practice. The story can be a bit different when we have `jsonb` field but then it can just handled through an instance based on a handwritten codec ;-)

On the other hand when you think about libraries like `selda` and different backends (mysql, postgresql etc.) this approach can be problematic. What is even worst we have not found any simple solution which would be easy to use etc.
I think that I can move this discussion there if you like as this is broader question about readability and extensibility of its API really bothers us.

**Comment by JordanMartinez** [comment]
> Usually types which we dealing with are simple and this encoding / decoding + standard newtype wrapping strategy seems to be working fine in practice.

I disagree. There's a difference between "being forced to use" the type classes `ToSqlRows`/`FromSqlRows` by the authors of a library and "being encouraged to use" those type classes. An escape hatch can be provided to those who need something outside of the norm.

I feel that this library should, at the very least, export [`unsafeQuery`](https://github.com/rightfold/purescript-postgresql-client/blob/master/src/Database/PostgreSQL.purs#L230-L241), and at the most, create a version of the `execute` and `command` functions that don't use the type classes.

> I think that I can move this discussion there if you like as this is broader question about readability and extensibility of its API really bothers us.

I don't think this dicussion should be moved there as another person could decide to build another library on top of this.

**Comment by paluh** [comment]
> I disagree. There's a difference between "being forced to use" the type classes ToSqlRows/FromSqlRows by the authors of a library and "being encouraged to use" those type classes.

Do you disagree with the convenience of `newtype` usage in this scenario? Could you provide any really problematic use case / scenario?

We both agree that this library enforces a user to use type classes and I never stated anything different I think.

On the other hand as a side note I can repeat that I don't really like this approach for the general encoding / decoding strategy. You can see my similar concerns in discussions like here: https://github.com/cprussin/purescript-httpure/pull/104#issuecomment-419585619 where I've proposed changes to HTTPure. I've also decided to move away from hyper exactly for its extensive usage of ad hoc polymorphism. Sorry for this off topic but I'm really want to proof that it is not my design rule etc.

Going back to the discussion:

* We can expose `unsafeQuery` but this nearly removes the main purpose of this library (a `Query` type).

* Do you have any proposition of an usable API which allows user to build queries and provide typed values and receive responses without type classes?
We can provide this kind of general API by replacing typeclasses with functions (our type classes are  single method dictionaries):

```purescript
query
    :: forall i o
    . (i -> Array Foreign)
    -> (Array Foreign -> o)
    -> Connection
    -> Query i o
    -> i
    -> Aff (Either PGError (Array o))
query conn (Query sql) values = ...
```
I'm not sure if the above is really usable in practice. It also removes quite nice type level API for row parsing which can be done either for `Row*` or just nested tuples:

https://github.com/rightfold/purescript-postgresql-client/blob/master/src/Database/PostgreSQL/Row.purs

Do you have any other `API` change proposition which removes the necessity of type classes usage but also provides and easy way to use and extend?


> I don't think this dicussion should be moved

Ok. I'm leaving this discussion here. I just think that API design should also follow broader view provided by libraries like selda.
@Kamirus - would you like to add something to this discussion from the "`purescript-selda` author" perspective?

**Comment by JordanMartinez** [comment]
Ok, so I'm realizing that my above comment was motivated by a different though very similar issue:  this library's inability to let me run a query and _not_ decode the result (i.e. I just want to print out the results to the console or something).

For example, a few weeks ago I had set up my queries using `selda` and was getting back a PGError. I had absolutely no idea why. So, I started running various debugging queries from SO and started tracking down the problem. However, `selda` requires me to fully specify the table I was querying. This library requires me to specify the exact number of columns the query returns via nested tuples. In my situation, I didn't care about decoding the result. I just wanted to print out the results of the query. Due to being unfamiliar with Postgres still, I wasn't always sure how many nested tuples I needed to run a query like `SELECT * FROM tableName;`
What should have been a rather trivial debugging session turned into one or two hours of debugging. 
I did finally figure out what the problem was. Due to copying and pasting some example code from another file that I modified in the file I wanted to use, I was importing the other file's database config value, not the one specified in the file I was using. 

If this library did expose the `unsafeQuery`, I could have bypassed most of the problems I experienced to do exactly that: unsafe queries.

> Do you disagree with the convenience of newtype usage in this scenario? Could you provide any really problematic use case / scenario?

Not necessarily. It's more of wanting this library to allow one to choose which approach they want to use--the type class approach or the SYTC approach--rather than restrict them to using only one, especially when such a change doesn't seem hard to make or maintain. Most of the time, I'll be using `selda` / the type class approach. 

> We can expose unsafeQuery but this nearly removes the main purpose of this library (a Query type).

I thought the main purpose of this library was two-fold:
- create PS bindings to `node-postgres-client` (primary)
- make it easier to encode/decode queries via the type classes (secondary)

> Ok. I'm leaving this discussion here. I just think that API design should also follow broader view provided by libraries like selda.

I think the heart of what I'm saying is, "99% of the time, I'll be using `selda` to handle queries because of its type-safe nature and ease of syntax. However, I don't think this library's design should force me to reimplement part of the library just so I can run an unsafe query every now and then."

**Comment by Kamirus** [comment]
IMO this library should provide these basic, unsafe functions that work on `Foreign`.

Right now selda uses these nested tuples as an intermediate representation because pgclient requires it.
Maybe it would be easier to implement the conversion directly between `Foreign` and records, maybe.
